pkg/repository/mirror.go

```diff
@@ -249,6 +326,71 @@ func (l *httpMirror) prepareURL(resource string) string {
 	return url
 }
 
+// Introduce implements the model.Model interface
+func (l *httpMirror) Introduce(id, name string, key *v1manifest.KeyInfo) error {
+	return errors.Errorf("introduce a fresher via the internet is not allowd, please set you mirror to a local one")
+}
+
+// Publish implements the model.Model interface
+func (l *httpMirror) Publish(manifest *v1manifest.Manifest, info model.ComponentInfo) error {
+	sid := uuid.New().String()
+
+	if info.Filename() != "" {
+		tarAddr := fmt.Sprintf("%s/api/v1/tarball/%s", l.Source(), sid)
+		resp, err := utils.PostFile(info, tarAddr, "file", info.Filename())
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+
+		if resp.StatusCode >= 300 {
+			return errors.Errorf("error on uplaod tarbal, server returns %d", resp.StatusCode)
+		}
+	}
+
+	payload, err := json.Marshal(manifest)
+	if err != nil {
+		return err
+	}
+	bodyBuf := bytes.NewBuffer(payload)
+	qpairs := []string{}
+	if info.Yanked() != nil {
+		qpairs = append(qpairs, fmt.Sprintf("%s=%t", "yanked", *info.Yanked()))
+	}
+	if info.Standalone() != nil {
+		qpairs = append(qpairs, fmt.Sprintf("%s=%t", "standalone", *info.Standalone()))
+	}
+	if info.Hidden() != nil {
+		qpairs = append(qpairs, fmt.Sprintf("%s=%t", "hidden", *info.Hidden()))
+	}
+	qstr := ""
+	if len(qpairs) > 0 {
+		qstr = "?" + strings.Join(qpairs, "&")
+	}
+	manifestAddr := fmt.Sprintf("%s/api/v1/component/%s/%s%s", l.Source(), sid, manifest.Signed.(*v1manifest.Component).ID, qstr)
+
+	resp, err := http.Post(manifestAddr, "text/json", bodyBuf)
```

Comment:
The default HTTP client has no timeout set, we should create a client with proper timeout setting and use it for requests.